File: src/main/java/team4/footwithme/vote/repository/CustomChoiceRepositoryImpl.java
package team4.footwithme.vote.repository;

import com.querydsl.jpa.impl.JPAQueryFactory;
import team4.footwithme.vote.domain.Choice;

import java.util.List;

import static team4.footwithme.vote.domain.QChoice.choice;
import static team4.footwithme.vote.domain.QVote.vote;
import static team4.footwithme.vote.domain.QVoteItem.voteItem;

public class CustomChoiceRepositoryImpl implements CustomChoiceRepository {

    private final JPAQueryFactory queryFactory;

    public CustomChoiceRepositoryImpl(JPAQueryFactory queryFactory) {
        this.queryFactory = queryFactory;
    }

    @Override
    public Long countByVoteItemId(Long voteItemId) {
        return queryFactory.select(choice.count())
            .from(choice)
            .where(choice.voteItemId.eq(voteItemId))
            .fetchOne();
    }

    @Override
    public List<Choice> findByMemberIdAndVoteId(Long memberId, Long voteId) {
        return queryFactory.select(choice)
            .from(choice)
            .join(voteItem).on(choice.voteItemId.eq(voteItem.voteItemId))
            .join(voteItem.vote, vote)
            .where(choice.memberId.eq(memberId)
                .and(vote.voteId.eq(voteId)))
            .fetch();
    }

    @Override
    public List<Long> findMemberIdsByVoteItemId(Long voteItemId) {
        return queryFactory.select(choice.memberId)
            .from(choice)
            .where(choice.voteItemId.eq(voteItemId))
            .fetch();
    }

    @Override
    public Long maxChoiceCountByVoteId(Long voteId) {
        return queryFactory.select(choice.voteItemId)
            .from(choice)
            .join(voteItem).on(choice.voteItemId.eq(voteItem.voteItemId))
            .join(voteItem.vote, vote)
            .where(vote.voteId.eq(voteId))
            .groupBy(choice.voteItemId)
            .orderBy(choice.count().desc())
            .fetchFirst();
    }
}

File: src/main/java/team4/footwithme/vote/repository/CustomChoiceRepositoryImpl.kt
package team4.footwithme.vote.repository

import com.querydsl.jpa.impl.JPAQueryFactory
import team4.footwithme.vote.domain.Choice
import team4.footwithme.vote.domain.QChoice
import team4.footwithme.vote.domain.QVote
import team4.footwithme.vote.domain.QVoteItem

class CustomChoiceRepositoryImpl(private val queryFactory: JPAQueryFactory) : CustomChoiceRepository {
    override fun countByVoteItemId(voteItemId: Long?): Long? {
        return queryFactory.select(QChoice.choice.count())
            .from(QChoice.choice)
            .where(QChoice.choice.voteItemId.eq(voteItemId))
            .fetchOne()
    }

    override fun findByMemberIdAndVoteId(memberId: Long?, voteId: Long?): List<Choice>? {
        return queryFactory.select(QChoice.choice)
            .from(QChoice.choice)
            .join(QVoteItem.voteItem).on(QChoice.choice.voteItemId.eq(QVoteItem.voteItem.voteItemId))
            .join(QVoteItem.voteItem.vote, QVote.vote)
            .where(
                QChoice.choice.memberId.eq(memberId)
                    .and(QVote.vote.voteId.eq(voteId))
            )
            .fetch()
    }

    override fun findMemberIdsByVoteItemId(voteItemId: Long?): List<Long>? {
        return queryFactory.select(QChoice.choice.memberId)
            .from(QChoice.choice)
            .where(QChoice.choice.voteItemId.eq(voteItemId))
            .fetch()
    }

    override fun maxChoiceCountByVoteId(voteId: Long?): Long {
        return queryFactory.select(QChoice.choice.voteItemId)
            .from(QChoice.choice)
            .join(QVoteItem.voteItem).on(QChoice.choice.voteItemId.eq(QVoteItem.voteItem.voteItemId))
            .join(QVoteItem.voteItem.vote, QVote.vote)
            .where(QVote.vote.voteId.eq(voteId))
            .groupBy(QChoice.choice.voteItemId)
            .orderBy(QChoice.choice.count().desc())
            .fetchFirst()!!
    }
}